### PR TITLE
DOC, TST: Wrapping of Keras models

### PR DESCRIPTION
**Pull request by stsievert**
**What does this PR implement?**
It documents and tests wrapping Keras models.

**Reference issues/PRs**
This inherits from #699.

### CODE REVIEW

**Comment by stsievert** [comment]
I've got this mostly working. The search can finish successfully, but not consistently. It might log an error on deserialization, or it might raise that error. Here are the three cases I've observed:

* The search finishes successfully, and not messages are printed/logged.
* A deserialization error happens and is logged, and then search finishes successfully. This starts [in distributed's _deserialize][2] or with `deepcopy(model)`, and ends in Tensorflow's [func_graph.py#L418][1]: "TypeError: NoneType is not iterable."
* The error logged above is raised.


[1]:https://github.com/tensorflow/tensorflow/blob/64dbe97a4d5a9a2cc69cacae7762157cd201a1d2/tensorflow/python/framework/func_graph.py#L418
[2]:https://github.com/dask/distributed/blob/ccc9d2a1d5c2ee8a0f37f146b48f64cffa5c373e/distributed/protocol/core.py#L131


<details>
<summary>Full example:</summary>

``` python
from dask.distributed import Client
from scipy.stats import loguniform
from sklearn.datasets import make_regression
import pickle

import tensorflow as tf
from tensorflow.keras.datasets import mnist as keras_mnist
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.utils import to_categorical
from scikeras.wrappers import KerasClassifier

from dask_ml.model_selection import IncrementalSearchCV
from sklearn.model_selection import RandomizedSearchCV


def mnist():
    (X_train, y_train), _ = keras_mnist.load_data()
    X_train = X_train[:100]
    y_train = y_train[:100]
    X_train = X_train.reshape(X_train.shape[0], 784)
    X_train = X_train.astype("float32")
    X_train /= 255
    Y_train = to_categorical(y_train, 10)
    return X_train, y_train


def _keras_build_fn(lr=0.01):
    layers = [
        Dense(512, input_shape=(784,), activation="relu"),
        Dense(10, input_shape=(512,), activation="softmax"),
    ]
    # See https://github.com/adriangb/scikeras/issues/24
    try:
        model = Sequential(layers)
    except (TypeError, RuntimeError):
        model = Sequential(layers)

    opt = tf.keras.optimizers.SGD(learning_rate=lr)
    model.compile(loss="categorical_crossentropy", optimizer=opt, metrics=["accuracy"])
    return model


if __name__ == "__main__":
    # always works and sometimes prints tracebacks (not often)
    #  c = Client(processes=False)

    # always works and no tracebacks printed
    #  c = Client(processes=False, n_workers=4, threads_per_worker=1)

    # fails
    c = Client(n_workers=4, threads_per_worker=1, processes=True)

    # Works most of the time. Sometimes prints tracebacks.
    #  c = Client(n_workers=2, threads_per_worker=1, processes=True)

    X, y = mnist()
    model = KerasClassifier(build_fn=_keras_build_fn, epochs=1, lr=0.1)
    params = {"lr": loguniform(1e-3, 1e-1)}

    search = IncrementalSearchCV(model, params, max_iter=3, decay_rate=None)
    search.fit(X, y, epochs=1, n_jobs=-1)
    assert search.best_score_ >= 0
```


</details>

**Comment by stsievert** [comment]
https://github.com/tensorflow/tensorflow/issues/35813 has the same issue. This issue might be resolved by putting the model creation in a lock: https://github.com/tensorflow/tensorflow/issues/35813#issuecomment-661763351

File: dask_ml/wrappers.py [hunk 1]
```diff
@@ -551,3 +551,58 @@ def _predict_proba(part, estimator):
...
+            r = super()._build_keras_model(*args, **kwargs)
+        return r
+
+
+class KerasClassifier(_BaseKerasWrapper, SciKerasClassifier):
+    pass
+
+
+class KerasRegressor(_BaseKerasWrapper, SciKerasRegressor):
+    pass
```
Comment (mrocklin) [inline dask_ml/wrappers.py:608]: It looks like these aren't used.  Can we safely drop them?

Comment (stsievert) [reply dask_ml/wrappers.py]: Yup, we can.

This is new from exploration in https://github.com/adriangb/scikeras/issues/24. Now, Tensorflow >= 2.3.0 is required to use TF with Dask-ML.

**Comment by stsievert** [comment]
This PR is ready for review again.

This PR is working now with the release of Tensorflow v2.3.0, released 3 days ago. I've removed the implementation of Dask-ML-specific wrappers and have detailed the install directions in the docs.

File: tests/model_selection/test_keras.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    ]
+
+    model = Sequential(layers)
+
+    opt = tf.keras.optimizers.SGD(learning_rate=lr)
+    model.compile(loss="categorical_crossentropy", optimizer=opt, metrics=["accuracy"])
+    return model
+
+
+@gen_cluster(client=True)
```
Comment (mrocklin) [inline tests/model_selection/test_keras.py:53]:
```suggestion
@gen_cluster(client=True, Worker=Nanny)
```

I'm curious, does Keras still work if we have to cross a process boundary?  We've had issues with serialization and Keras in the past.

**Comment by mrocklin** [comment]
+1 generally.

I added a small question about serialization, but that can be handled separately from this PR.

File: tests/model_selection/test_keras.py (hunk 2 above)
Comment (stsievert) [reply tests/model_selection/test_keras.py]: Yes, at least in my testing with `LocalCluster(processes=True, n_workers=5, threads_per_worker=2)`.

This is new behavior: Dask-ML requires Tensorflow v2.3.0, which is only 3 days old.

**Comment by adriangb** [comment]
Sorry if it is too late for this, but would it be possible to introduce using the [subclassing interface](https://github.com/adriangb/scikeras#subclassing-wrappers) instead? That interface is much more robust than the `build_fn` interface in my opinion.

**Comment by stsievert** [comment]
> the subclassing interface ... is much more robust than the build_fn interface in my opinion.

Robust how? How does that differ from the function initialization I currently have implemented?

**Comment by adriangb** [comment]
> Robust how? How does that differ from the function initialization I currently have implemented?

It's more compliant with the Scikit-Learn API.  Some internal parts of Scikit-Learn may inspect the `__init__` signature of the model instance and get upset if they find an attribute (ex: `verbose`) that was set during instantiation but was not part of the `__intit_` signature of `BaseWrapper`.

**Comment by stsievert** [comment]
I can add a note about using the subclassing interface if they run into any unexpected interfaces. But the function interface is a lot simpler for users, and the example I documented doesn't raise any errors.

File: docs/source/keras.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+   from tensorflow.keras.utils import to_categorical
+   import numpy as np
+   from typing import Tuple
+
+   def get_mnist() -> Tuple[np.ndarray, np.ndarray]:
+       (X_train, y_train), _ = mnist.load_data()
+       X_train = X_train.reshape(X_train.shape[0], 784)
+       X_train = X_train.astype("float32")
+       X_train /= 255
+       return X_train, y_train
```
Comment (TomAugspurger) [inline docs/source/keras.rst:95]: Does this also work with Dask Arrays?

File: ci/posix.yaml [hunk 4]
```diff
@@ -40,6 +41,11 @@ jobs:
     displayName: "install PyTorch"
     condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
 
+  - bash: |
+      pip install tensorflow>=2.3.0 scikeras>=0.1.8
+    displayName: "install Tensorflow and SciKeras"
+    # condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```
Comment (TomAugspurger) [inline ci/posix.yaml:47]:
```suggestion
    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```

**Comment by TomAugspurger** [comment]
Hmm actually, can you double check the CI? https://dev.azure.com/dask-dev/dask/_build/results?buildId=1432&view=logs&j=d699f7f7-fbb1-5fe7-7f49-16a980a27dcf&t=47702d4b-eb38-52c1-7644-7bfb10716412&l=906 says that the keras tests are skipped. The "Install tensorflow" output section is blank.

File: ci/posix.yaml [hunk 5]
```diff
@@ -44,7 +44,7 @@ jobs:
   - bash: |
       pip install tensorflow>=2.3.0 scikeras>=0.1.8
     displayName: "install Tensorflow and SciKeras"
-    # condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
+    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```
Comment (TomAugspurger) [inline ci/posix.yaml:47]:
```suggestion
    # condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```

File: docs/source/keras.rst (hunk 3 above)
Comment (stsievert) [reply docs/source/keras.rst:95]: I presume so. `X_train` and `y_train` are NumPy arrays.

It looks like SciKeras accepts NumPy "array-like" objects, which includes NumPy arrays, Dataframes, and sparse arrays: [wrappers.py#L366][1]

[1]: https://github.com/adriangb/scikeras/blob/48a727b91128e9f75f3cb848c7fea33bfef40659/scikeras/wrappers.py#L366.

Comment (adriangb) [reply docs/source/keras.rst:95]: Yep, by calling `check_array` ([line](https://github.com/adriangb/scikeras/blob/48a727b91128e9f75f3cb848c7fea33bfef40659/scikeras/wrappers.py#L389)) SciKeras handles the same data types as most Scikit-Learn models can handle.

**Comment by stsievert** [comment]
I've resolved the issue; now I use `python -c "import tensorflow as tf; print('tf.__version__')` to print the Tensorflow/Keras and PyTorch/Skorch versions to stdout. I saw (the correct versions) in 18466dd; I expect 8ec00fc to pass. After that, I think this PR is ready for merge after the condition in `ci/posix.yaml` is un-commented.

**Comment by stsievert** [comment]
This PR is nearly ready for merge.

However, I'm getting some `TimeoutError`s on `test_keras`. I can reproduce these errors with pytest but can't reproduce them with `Client(LocalCluster(n_workers=2, threads_per_worker=1))`. These tests appear deterministic locally under Python 3.8.4 but do not appear deterministic on Travis CI (linux37 and linux-earliest failed in 7a29030, but ony linux-earliest failed in 7ace25d).

I'm not sure how to resolve this. Any tips?

**Comment by mrocklin** [comment]
It looks like it's downloading data perhaps?  Maybe that is what causes it to take a while?

```
>               result = await future
E               tornado.util.TimeoutError: Timeout

/usr/share/miniconda/envs/dask-ml-test/lib/python3.6/site-packages/distributed/utils_test.py:920: TimeoutError
----------------------------- Captured stdout call -----------------------------
Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/mnist.npz

    8192/11490434 [..............................] - ETA: 0s
 5562368/11490434 [=============>................] - ETA: 0s
11493376/11490434 [==============================] - 0s 0us/step
----------------------------- Captured stderr call -----------------------------
```

File: tests/model_selection/test_keras.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            version.parse(scikeras.__version__) < version.parse("0.1.8"),
+            reason="partial_fit support",
+        ),
+    ]
+except ImportError:
+    pytestmark = pytest.mark.skip(reason="Missing tensorflow or scikeras")
+
+
+def mnist() -> Tuple[np.ndarray, np.ndarray]:
+    (X_train, y_train), _ = keras_mnist.load_data()
```
Comment (mrocklin) [inline tests/model_selection/test_keras.py:36]: Maybe we shouldn't download mnist as part of the test suite.  Maybe there is some smaller testing dataset that we can play with instead?

Comment (stsievert) [reply tests/model_selection/test_keras.py]: Thanks!

Comment (stsievert) [reply tests/model_selection/test_keras.py]: Now, I mirror the MNIST dataset.

File: tests/model_selection/test_keras.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+from scipy.stats import loguniform
+from sklearn.datasets import make_classification
+
+from dask_ml.model_selection import IncrementalSearchCV
+
+try:
+    import scikeras
+    import tensorflow as tf
+    from scikeras.wrappers import KerasClassifier
+    from tensorflow.keras.datasets import mnist as keras_mnist
```
Comment (mrocklin) [inline tests/model_selection/test_keras.py:18]:
```suggestion
```

**Comment by TomAugspurger** [comment]
Seems to have worked, right @stsievert? You'll just need to disable for PRs now.

**Comment by stsievert** [comment]
I think this PR is good, past the `TimeoutError`s error on the model selection tests.

File: ci/posix.yaml [hunk 8]
```diff
@@ -37,9 +38,20 @@ jobs:
...
     condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
 
+  - bash: |
+      source activate dask-ml-test
+      pip install tensorflow>=2.3.0
+      pip install scikeras>=0.1.8
+      python -c "import tensorflow as tf; print('TF ' + tf.__version__)"
+      python -c "import scikeras; print('SciKeras ' + scikeras.__version__)"
+    displayName: "install Tensorflow and SciKeras"
+    # condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```
Comment (TomAugspurger) [inline ci/posix.yaml:53]:
```suggestion
    # condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```

**Comment by stsievert** [comment]
All checks are passing, at least on this branch. I think we'll still get some TimeoutErrors on the master branch after 1baa382. Hopefully these are resolved in #527.

**Comment by TomAugspurger** [comment]
Thanks!

**Comment by stsievert** [comment]
Yup, both PyTorch and Keras tests are running in the latest CI job on master: https://dev.azure.com/dask-dev/dask/_build/results?buildId=1498&view=logs&j=ec4c5d99-1a39-52ae-ac68-aabdfb55b8a0

**Comment by TomAugspurger** [comment]
Great, thanks.

**Pull request merged**